calorie related math effect

GuardianDll opened the pull request:
#### Summary
None
#### Purpose of change
Tried to add a way for math to read item calorie content, decided the best approach would be moving entire dialogue condition to the math
#### Describe the solution
make a math function `calories()`, that contain `stored_kcal` (as `calories()` or `calories('format': 'raw')`), stored_kcal_percentage (as `calories('format': 'percent')`) and also can read the caloric content of the item, if one of the talkers is item
#### Testing
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/67688115/d6640570-86b9-407d-bf6d-0550c39a968c)
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/67688115/6a65659c-d866-4cad-9384-ae69b947fcd4)
#### Additional context
~~Albeit wonky, it worked before, but then i was told if i want optional options, i need to use kwargs instead of params, and when i changed it, it stopped working with items, which was kinda the point of all of it~~ it worked, i just didn't use the proper EoC, duh

GuardianDll committed:
Add `calories()` math function

GuardianDll committed:
add assigning function

GuardianDll committed:
remove u_val version of the kcal checks

GuardianDll committed:
replaces all instances of u_val('kcal') with new function

GuardianDll committed:
document change

GuardianDll committed:
fix clang tidy error

mqrause commented on src/math_parser_diag.cpp:
```diff
@@ -1317,6 +1317,56 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
+                npc dummy;
+                return dummy.compute_effective_nutrients( *it->get_item() ).kcal();
+            }
+        } else {
+            debugmsg( "For calories(), talker is not character nor item" );
+            return 0;
+        };
```
```suggestion
        }
        
        debugmsg( "For calories(), talker is not character nor item" );
        return 0;
```

This should fix the issue. If it's wrapped in an `else`, it won't catch paths like when `if( it && *it )` was false.

GuardianDll committed:
Update src/math_parser_diag.cpp

Co-authored-by: mqrause <38702195+mqrause@users.noreply.github.com>

GuardianDll committed:
Update src/math_parser_diag.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

GuardianDll committed:
resolve the rest of shiv, fix (potentially) few tests

GuardianDll committed:
missed the zero

GuardianDll committed:
remove percent calories assigning since you can't assign calories as percent anymore

GuardianDll committed:
remove math kcal test because it's not a shim

GuardianDll committed:
the way this test is made drive me insane

GuardianDll committed:
i was told this can fix the test

GuardianDll committed:
spare me

GuardianDll committed:
now this one is my fault

GuardianDll committed:
Merge branch 'CleverRaven:master' into calorie_math

GuardianDll committed:
fix MoM using old syntax

anothersimulacrum commented on src/math_parser_diag.cpp:
```diff
@@ -1321,6 +1321,55 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
+                int divisor = d.actor( beta )->get_healthy_kcal() / 100;
+                //if no data, default to default height of 175cm
+                if( divisor == 0 ) {
+                    divisor = 118169 / 100;
```
Where does the 118169 magic number come from? Either use the calculation that generates it or just have a debugmsg here and give some safe value, not a (liable to change) calculated value.

GuardianDll replied:
right from the original `stored_kcal_percentage` implementation
https://github.com/CleverRaven/Cataclysm-DDA/pull/73136/files#diff-534f433e7a0b5ec5c233390037e131f3717e8aa5f347c47e650a65a1ddc95279L2133

anothersimulacrum replied:
Sorry, that was a rhetorical question. There shouldn't be a magic number like this.

anothersimulacrum commented on src/math_parser_diag.cpp:
```diff
@@ -1321,6 +1321,55 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
+
+        if( format == "percent" ) {
+            if( d.actor( beta )->get_character() ) {
+                int divisor = d.actor( beta )->get_healthy_kcal() / 100;
```
Do floating point math with this, all EOC math is.
```suggestion
                double divisor = d.actor( beta )->get_healthy_kcal() / 100.0;
```

GuardianDll committed:
Do floating point math

Co-authored-by: anothersimulacrum <anothersimulacrum@gmail.com>

GuardianDll committed:
get rid of magic number

GuardianDll committed:
revert `Do floating point math` because i don't want static cast all the returns in the function.
better approach would be divide this function in two, one for raw value that return int, one for percent value that return double
or function that return healthy kcal directly

GuardianDll committed:
add `-> double`

anothersimulacrum commented on src/math_parser_diag.cpp:
```diff
@@ -1321,6 +1321,59 @@ std::function<double( dialogue & )> vision_range_eval( char scope,
+        if( format == "percent" )
+        {
+            if( d.actor( beta )->get_character() ) {
+                int divisor = d.actor( beta )->get_healthy_kcal() / 100.0;
```
```suggestion
                double divisor = d.actor( beta )->get_healthy_kcal() / 100.0;
```

GuardianDll replied:
right, it needs to be turned back now

GuardianDll committed:
Update src/math_parser_diag.cpp

Co-authored-by: anothersimulacrum <anothersimulacrum@gmail.com>
